### DISCUSSION TITLE
ISR with AWS ECS fargate

### DISCUSSION THREAD

**Discussion by RishikeshDarandale** [vercel/next.js#16852] [Help]
Hi,
Can anybody help me on serving the `ISR` feature of next.js with AWS ECS fargate service? Basically, ISR works on SWR principle, thus storing the next.js build local to container will not work. 

AWS EFS can be one option, but how to use it with it?
A common mount point with EFS to Fargate will do, but running the `npm run build` simultaneously on common mount point for `build` directory make sense?

**EDIT**: I tried with EFS mount point with CMD of docker container as `npm run build && npm run start`, but with two container I could see the stale page twice.

Thanks

**Comment by ash1425** [comment]
Hello @RishikeshDarandale 
You should also add [this](https://nextjs.org/docs/basic-features/data-fetching/incremental-static-regeneration#self-hosting-isr) configuration as mentioned below to ensure that in-memory cache is not used.

```
module.exports = {
  experimental: {
    // Defaults to 50MB
    isrMemoryCacheSize: 0,
  },
}
```


With above - I could test it on local k8s using shared file system mount for .next folder.

Could you please also share how did you configure it on aws using EFS ? When did you copy .next to EFS ?

**Reply by RishikeshDarandale** [reply]
@ash1425 , this was asked about 2 years back when this experimental configuration was not available. I tried attaching EFS mount point to ECS fargate with two tasks, but it didn't worked that time. Thus, I had to avoid the ISR in multi task ECS fargate environment.

I am just thinking aloud there, the build pipeline needs to solid when coping the build onto a common mount point.

**Reply by RishikeshDarandale** [reply]
Currently, I am using the workaround suggested [here](https://github.com/vercel/next.js/discussions/16733). Use SSR with CDN.

**Reply by ash1425** [reply]
yes, I am trying to get my head around copying the build artefacts to shared EFS, also rollback is little difficult in this case.

I am planning to get rid of ISR for now and SSG for few locales only instead of building for all 80 locales statically. This way we will need deployment for each change, but would have consistent pages across instances.

**Reply by RishikeshDarandale** [reply]
If you are using CDN like (cloudflare / fastly), then you can use [stale-while-revalidate](https://developer.fastly.com/learning/concepts/stale/#stale-while-revalidate-eliminate-origin-latency) directive to achieve the same concept of ISR using SSR. (FYI: [cloudfront](https://repost.aws/questions/QUBHVmYXJ5RhG1ZCdAySc3MQ/cloudfront-support-for-stale-while-revalidate) does not support this directive!)

- SSR page will respond with `Cache-Control` header say with freshness of the page as 6 hours with revalidate 5 mins.
- Once the page is cached at CDN, it will be fresh for 6 hours.
- If any subsequent requests comes in after 6 hours and within 5 mins, stale copy is served from CDN. At the same time CDN will hit origin to get a new copy and update the cache.
- If no requests for the same page comes in 5 minutes after the freshness, then it will be a hit to origin to get a new copy.

**Reply by ash1425** [reply]
Thank you. We use cloudfront as CDN and stale-while-revalidate wont help :( 

We are planning to move to SSR + Cloudfront

**Reply by marcesengel** [reply]
@RishikeshDarandale @ash1425 a little late to the party, but we're planning to do this using a multi-stage docker build and relatively new `standalone` mode. When deploying we won't only change the ECS docker image but also swap the mounted EFS (which will be mounted to `/app/.next/server/pages`) with the task configuration update in order to have clean cuts between versions.

I'll try to let you know the result, any tips/advice beforehand? Thanks!

**Reply by marcesengel** [reply]
After some more digging we've also decided to pivot towards SSR + CloudFront as the upsides of SSG with ISR don't really weigh up the additional complexity in deployment and introducing a new possible bottleneck (EFS) to our clusters 😆

**Comment by taraspos** [comment]
Sharing solution I came up with, which is working well enough for me:

1. Create new cache handler based on the default [incremental-cache/file-system-cache](https://github.com/vercel/next.js/blob/v15.1.4/packages/next/src/server/lib/incremental-cache/file-system-cache.ts)
   - `cache_handlers/fs.js`:
      ```js
      const FileSystemCache = require("next/dist/server/lib/incremental-cache/file-system-cache").default;
      const path = require('path');
      
      // This allows us to re-use official file-system-cache implementation,
      // but with a custom serverDistDir, so we can use it to store shared ISR cache
      // between multiple copies of the app.
      class CustomFileSystemCache extends FileSystemCache {
          constructor(ctx) {
              super(ctx);
              if (process.env.NEXT_SHARED_CACHE_LOCATION) {
                  this.serverDistDir = path.resolve(process.env.NEXT_SHARED_CACHE_LOCATION);
                  if (process.env.NEXT_DEPLOYMENT_ID) {
                      // append deployment id to the cache location
                      // to avoid conflicts between different deployments
                      // when using EFS, this will leave bunch of unsed files after every deployment
                      // you will need to clean them up manually at some point
                      // you can configure automatic transition to IA/Archived to save cost
                      this.serverDistDir += '_' + process.env.NEXT_DEPLOYMENT_ID;
                  }
              }
          }
      }
      
      module.exports = CustomFileSystemCache;
      ```
    - References:
      - https://nextjs.org/docs/app/building-your-application/deploying#caching-and-isr
      - https://nextjs.org/docs/app/building-your-application/deploying#version-skew

2. Configure your app to use your custom cache handler and not to use memory cache:
    ```js
    const nextConfig = {
      cacheHandler: path.resolve('./cache_handlers/fs.js'),
      cacheMaxMemorySize:  0,
    ...
    ```
    - https://nextjs.org/docs/app/api-reference/config/next-config-js/incrementalCacheHandlerPath

3. Create EFS access point like:
    ```HCL
    resource "aws_efs_access_point" "mountpoint" {  
      file_system_id = var.efs_id
      root_directory {
        path = "/nextjs/${var.app_name}/cache"
        creation_info {
          owner_uid   = 1001
          owner_gid   = 1001
          permissions = "755"
        }
      }
    }
    ```

4. Configure your application with:
  - (optional) Set `NEXT_DEPLOYMENT_ID` during build time to add additional version skew protection
  - Set `NEXT_SHARED_CACHE_LOCATION` (in runtime) to a location one level higher than FS mount point. This way some cache can be still shared between deployments as intended:
    - https://github.com/vercel/next.js/blob/48f2588b0fea2bffb5bf6534169ee112438786a6/packages/next/src/server/lib/incremental-cache/file-system-cache.ts#L434-L443
